<a href="https://colab.research.google.com/github/YaninaK/anomaly-detection/blob/b1/notebooks/01_EDA_%26_data_preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Анализ аномалий в начислениях за тепловую энергию


## 1. Анализ и подготовка данных



In [1]:
initiate = False
if initiate:
  !git init -q
  !git clone -b b1  https://github.com/YaninaK/anomaly-detection.git -q

  from google.colab import drive
  drive.mount('/content/drive')

  !unzip -u -q /content/drive/MyDrive/ML_projects/08_anomaly_detection/data/01_raw/task#3.zip -d /content/anomaly-detection/data/01_raw

%cd /content/anomaly-detection/notebooks

/content/anomaly-detection/notebooks


In [2]:
import os
import sys

sys.path.append(os.getcwd())
sys.path.append(os.path.join(os.getcwd(), "..", "src", "anomaly_detection"))

In [3]:
import datetime

import numpy as np
import pandas as pd

from data.make_dataset import load_data
from data.preprocess import Preprocess

In [4]:
import warnings
warnings.filterwarnings('ignore')

In [5]:
PATH = "/content/anomaly-detection/"

## 1. Чтение данных

In [6]:
folder_path = '../data/01_raw/'
file_names = os.listdir(folder_path)

file_names

['Апрель 2022 Уфа.xls',
 'Ноябрь 2022 Уфа.xlsx',
 'Июнь 2022 Уфа.xls',
 'Июнь 2023 Уфа.xlsx',
 'Сентябрь 2021 Уфа.xls',
 'Температуры, продолжительность ОП.xls',
 'Февраль 2022 Уфа.xls',
 'Январь 2023 Уфа.xlsx',
 'Январь 2022 Уфа.xls',
 'Май 2023 Уфа.xlsx',
 'Сентябрь 2022 Уфа.xlsx',
 'Июль 2022 Уфа.xlsx',
 'Тип строения, этажность, площадь, год постройки.xlsx',
 'Декабрь 2022 Уфа.xlsx',
 'Примеры аномалий.docx',
 'Ноябрь 2021 Уфа.xls',
 'Август 2021 Уфа.xls',
 'Март 2023 Уфа.xlsx',
 'Июль 2021 Уфа.xls',
 'Май 2022 Уфа.xls',
 'Декабрь 2021 Уфа.xls',
 'Октябрь 2021 Уфа.xls',
 'Октябрь 2022 Уфа.xlsx',
 'Март 2022 Уфа.xls',
 'Август 2022 Уфа.xlsx',
 'Февраль 2023 Уфа.xlsx',
 '.gitkeep',
 'Апрель 2023 Уфа.xlsx']

In [7]:
regenerate = False
data, temperature, buildings = load_data(folder_path, regenerate, path=PATH)

100%|██████████| 24/24 [00:14<00:00,  1.61it/s]


In [8]:
print(f"data.shape = {data.shape}\n")
data.sample(2)

data.shape = (75391, 9)



index Подразделение    № ОДПУ Вид энерг-а ГВС  \
61993    928           Уфа    801778             NaN   
40324   4226           Уфа  00117149             NaN   

                       Адрес объекта          Тип объекта  \
61993  г Уфа, ул. Орджоникидзе, д.22  Многоквартирный дом   
40324     г Уфа, ул. Глазовская, д.2      Другое строение   

      Дата текущего показания  Текущее потребление, Гкал Период потребления  
61993              2022-04-01                    208.156         2022-03-01  
40324                     NaT                        NaN         2021-11-01

In [9]:
data["Адрес объекта"].nunique()

4278

* Тип строения, этажность, площадь, год постройки

In [10]:
print(f"buildings.shape = {buildings.shape}\n")
buildings.sample(2)

buildings.shape = (5873, 5)



Адрес объекта          Тип Объекта  \
5404              г Уфа, ул. Цюрупы, д.110  Многоквартирный дом   
5090  г Уфа, ул. Транспортная, д.28 корп.4      Другое строение   

     Этажность объекта       Дата постройки  Общая площадь объекта  
5404                 5  1965-01-01 00:00:00                4633.00  
5090                 2                  NaN                   0.01

In [11]:
buildings["Адрес объекта"].nunique()

5695

* Температуры, продолжительность отопительного периода.

  * ОЗП — осенне-зимний период, время, когда работает система отопления.

In [12]:
temperature = pd.read_excel(
    f'{folder_path}Температуры, продолжительность ОП.xls', index_col=1, skiprows=1
).T[1:].reset_index()

temperature

Период      index  Тн.в, град.С  Продолжительность ОЗП, сут.
0      2021-07-01           NaN                          NaN
1      2021-08-01           NaN                          NaN
2      2021-09-01           NaN                          NaN
3      2021-10-01      6.696552                         29.0
4      2021-11-01     -1.000000                         31.0
5      2021-12-01     -5.176667                         30.0
6      2022-01-01    -11.500000                         31.0
7      2022-02-01     -6.735714                         28.0
8      2022-03-01     -5.151613                         31.0
9      2022-04-01      5.650000                         34.0
10     2022-05-01           NaN                          NaN
11     2022-06-01           NaN                          NaN
12     2022-07-01           NaN                          NaN
13     2022-08-01           NaN                          NaN
14     2022-09-01           NaN                          NaN
15     2022-10-01      7.137037                         27.0
16     2022-11-01     -1.606452                         31.0
17     2022-12-01    -10.903333                         30.0
18     2023-01-01    -12.745161                         31.0
19     2023-02-01     -8.944828                         29.0
20     2023-03-01     -2.266667                         30.0
21     2023-04-01      7.818750                         32.0
22     2023-05-01           NaN                          NaN
23     2023-06-01           NaN                          NaN

## 2. Анлиз и подготовка данных

In [13]:
preprocess = Preprocess()
data, buildings = preprocess.fit_transform(data, buildings)

### 2.1. Неуникальные адреса объектов

In [14]:
df = buildings.groupby(["Тип Объекта", "Адрес объекта"])["Общая площадь объекта"].count()
df = df[df > 1].reset_index()
df

Тип Объекта  \
0                      Другое строение   
1                      Другое строение   
2                      Другое строение   
3                      Другое строение   
4                      Другое строение   
5                      Другое строение   
6                      Другое строение   
7                      Другое строение   
8                      Другое строение   
9                      Другое строение   
10                     Другое строение   
11                     Другое строение   
12                     Другое строение   
13                     Другое строение   
14                     Другое строение   
15                     Другое строение   
16                     Другое строение   
17                     Другое строение   
18                     Другое строение   
19                     Другое строение   
20                     Другое строение   
21                     Другое строение   
22                     Другое строение   
23                     Другое строение   
24                     Другое строение   
25                     Другое строение   
26                     Другое строение   
27                     Другое строение   
28                     Другое строение   
29                     Другое строение   
30                     Другое строение   
31                     Другое строение   
32                     Другое строение   
33                     Другое строение   
34                     Другое строение   
35                     Другое строение   
36                     Другое строение   
37                     Другое строение   
38                     Другое строение   
39                     Другое строение   
40                     Другое строение   
41                     Другое строение   
42                     Другое строение   
43                     Другое строение   
44                     Другое строение   
45                     Другое строение   
46                 Многоквартирный дом   
47                 Многоквартирный дом   
48                 Многоквартирный дом   
49                 Многоквартирный дом   
50                 Многоквартирный дом   
51                 Многоквартирный дом   
52                 Многоквартирный дом   
53  Учебное заведение, комбинат, центр   
54                         Частный дом   
55                         Частный дом   
56                         Частный дом   
57                         Частный дом   

                                Адрес объекта  Общая площадь объекта  
0             г Уфа, б-р. Тухвата Янаби, д.34                      2  
1                 г Уфа, пр-кт Октября, д.153                      2  
2                  г Уфа, пр-кт Октября, д.31                      2  
3                  г Уфа, пр-кт Октября, д.91                      2  
4    г Уфа, ул. Ахметова, д.316 корп.а потери                      2  
5                   г Уфа, ул. Белякова, д.25                      2  
6                     г Уфа, ул. Войкова, д.1                      2  
7                г Уфа, ул. Вологодская, д.29                      2  
8     г Уфа, ул. Дмитрия Донского, д.1 корп.5                      2  
9     г Уфа, ул. Дмитрия Донского, д.5 корп.4                      2  
10      г Уфа, ул. Индустриальное шоссе, д.26                      3  
11              г Уфа, ул. Инициативная, д.12                      2  
12              г Уфа, ул. Карла Маркса, д.56                      2  
13       г Уфа, ул. Карла Маркса, д.62 корп.1                      2  
14                    г Уфа, ул. Кирова, д.27                      2  
15       г Уфа, ул. Комсомольская, д.1 корп.1                      2  
16     г Уфа, ул. Комсомольская, д.122 корп.Б                      3  
17             г Уфа, ул. Комсомольская, д.15                      2  
18             г Уфа, ул. Ленина, д.99 корп.2                      2  
19      г Уфа, ул. Лесотехникума, д.49 корп.1                      2  
20             г Уфа, ул. Лесотехникума, д.92            

In [15]:
df["Тип Объекта"].value_counts()

Тип Объекта
Другое строение                       46
Многоквартирный дом                    7
Частный дом                            4
Учебное заведение, комбинат, центр     1
Name: count, dtype: int64

#### 2.1.1 Объекты типа ```Другое строение```

In [16]:
s = "Другое строение"
selection = df[df["Тип Объекта"] == s]["Адрес объекта"].tolist()
n_address = len(selection)

cond_1a = data["Тип объекта"]==s
cond_2a = data["Адрес объекта"].isin(selection)
cond_1b = buildings["Тип Объекта"]==s
cond_2b = buildings["Адрес объекта"].isin(selection)

df1 = data[cond_1a & cond_2a]
df2 = buildings[cond_1b & cond_2b]

n_objects = df2.shape[0]
n_billed_address = df1["Адрес объекта"].nunique()
n_odpu = df1["№ ОДПУ"].nunique()
consumed = df1["Текущее потребление, Гкал"].sum()

no_bills = set(selection) - set(df1["Адрес объекта"])
no_bills_objects = df2[df2['Адрес объекта'].isin(no_bills)].iloc[:, :-1]

print(f"На {n_address} неуникальных адресах находится {n_objects} объектов типа {s}.")
print(f'Общая площадь этих объектов: {df2["Общая площадь объекта"].sum()}.')
print(f'На {n_billed_address} адреса из {n_address} приходится {n_odpu} № ОДПУ')
print(f'Oбщий объем потребления составил{consumed: .0f} Гкал.')
print(f'На {len(no_bills)} неуникальных адресах ({len(no_bills_objects)} объектов типа {s}) показаний ОДПУ нет.\n')

no_bills_objects

На 46 неуникальных адресах находится 106 объектов типа Другое строение.
Общая площадь этих объектов: 2.62.
На 34 адреса из 46 приходится 66 № ОДПУ
Oбщий объем потребления составил 75883 Гкал.
На 12 неуникальных адресах (29 объектов типа Другое строение) показаний ОДПУ нет.



Адрес объекта      Тип Объекта  \
1148  г Уфа, ул. Ахметова, д.316 корп.а потери  Другое строение   
1149  г Уфа, ул. Ахметова, д.316 корп.а потери  Другое строение   
1579                   г Уфа, ул. Войкова, д.1  Другое строение   
1580                   г Уфа, ул. Войкова, д.1  Другое строение   
2060     г Уфа, ул. Индустриальное шоссе, д.26  Другое строение   
2061     г Уфа, ул. Индустриальное шоссе, д.26  Другое строение   
2062     г Уфа, ул. Индустриальное шоссе, д.26  Другое строение   
2739    г Уфа, ул. Комсомольская, д.122 корп.Б  Другое строение   
2740    г Уфа, ул. Комсомольская, д.122 корп.Б  Другое строение   
2742    г Уфа, ул. Комсомольская, д.122 корп.Б  Другое строение   
2777            г Уфа, ул. Комсомольская, д.15  Другое строение   
2778            г Уфа, ул. Комсомольская, д.15  Другое строение   
3127            г Уфа, ул. Лесотехникума, д.92  Другое строение   
3128            г Уфа, ул. Лесотехникума, д.92  Другое строение   
3130     г Уфа, ул. Лесотехникума, д.92 корп.2  Другое строение   
3131     г Уфа, ул. Лесотехникума, д.92 корп.2  Другое строение   
4763                г Уфа, ул. Рязанская, д.14  Другое строение   
4764                г Уфа, ул. Рязанская, д.14  Другое строение   
5184                г Уфа, ул. Ульяновых, д.59  Другое строение   
5185                г Уфа, ул. Ульяновых, д.59  Другое строение   
5186                г Уфа, ул. Ульяновых, д.59  Другое строение   
5187                г Уфа, ул. Ульяновых, д.59  Другое строение   
5188                г Уфа, ул. Ульяновых, д.59  Другое строение   
5240    г Уфа, ул. Уфимское шоссе, д.24 корп.4  Другое строение   
5241    г Уфа, ул. Уфимское шоссе, д.24 корп.4  Другое строение   
5272    г Уфа, ул. Уфимское шоссе, д.38 корп.1  Другое строение   
5273    г Уфа, ул. Уфимское шоссе, д.38 корп.1  Другое строение   
5624            г Уфа, ул. Шафиева, д.8 корп.1  Другое строение   
5625            г Уфа, ул. Шафиева, д.8 корп.1  Другое строение   

      Этажность объекта Дата постройки  Общая площадь объекта  
1148                  0            NaT                   0.00  
1149                  0            NaT                   0.11  
1579                  0            NaT                   0.01  
1580                  1            NaT                   0.01  
2060                  1     1975-01-01                   0.01  
2061                  5     1975-01-01                   0.01  
2062                  5     1975-01-01                   0.03  
2739                  0            NaT                   0.01  
2740                  0            NaT                   0.13  
2742                  3            NaT                   0.01  
2777                  0     2013-01-01                   0.01  
2778                  0     2013-01-01                   0.01  
3127                  1            NaT                   0.01  
3128                  3            NaT                   0.04  
3130                  1            NaT                   0.01  
3131                  4            NaT                   0.01  
4763                  1            NaT                   0.01  
4764                  5            NaT                   0.01  
5184                  1     1985-01-01                    NaN  
5185                  2     1985-01-01                    NaN  
5186                  3     1985-01-01                    NaN  
5187                  5     1985-01-01                   0.01  
5188                  5     1985-01-01                   0.01  
5240                  1            NaT                   0.01  
5241                  2            NaT                   0.01  
5272                  1            NaT                   0.01  
5273                  6            NaT                   0.04  
5624                  1     1981-01-01                   0.01  
5625                 12     1981-01-01                   0.01

#### 2.1.2 Объекты типа ```Многоквартирный дом```

In [17]:
s = "Многоквартирный дом"
selection = df[df["Тип Объекта"] == s]["Адрес объекта"].tolist()
n_address = len(selection)

cond_1a = data["Тип объекта"]==s
cond_2a = data["Адрес объекта"].isin(selection)
cond_1b = buildings["Тип Объекта"]==s
cond_2b = buildings["Адрес объекта"].isin(selection)

df1 = data[cond_1a & cond_2a]
df2 = buildings[cond_1b & cond_2b]

n_objects = df2.shape[0]
consumed = df1["Текущее потребление, Гкал"].sum()

no_bills = set(selection) - set(df1["Адрес объекта"])
no_bills_objects = df2[df2['Адрес объекта'].isin(no_bills)].iloc[:, :-1]

print(f"На {n_address} неуникальных адресах находится {n_objects} объектов типа {s}.")
print(f'Общая площадь этих объектов:{df2["Общая площадь объекта"].sum(): .0f}.')
print(f'Oбщий объем потребления составил{consumed: .0f} Гкал.\n')

df2

На 7 неуникальных адресах находится 14 объектов типа Многоквартирный дом.
Общая площадь этих объектов: 41929.
Oбщий объем потребления составил 9436 Гкал.



Адрес объекта          Тип Объекта  \
1583         г Уфа, ул. Вологодская, д.13  Многоквартирный дом   
1584         г Уфа, ул. Вологодская, д.13  Многоквартирный дом   
1603         г Уфа, ул. Вологодская, д.20  Многоквартирный дом   
1604         г Уфа, ул. Вологодская, д.20  Многоквартирный дом   
2101  г Уфа, ул. Интернациональная, д.113  Многоквартирный дом   
2102  г Уфа, ул. Интернациональная, д.113  Многоквартирный дом   
2397              г Уфа, ул. Кирова, д.95  Многоквартирный дом   
2398              г Уфа, ул. Кирова, д.95  Многоквартирный дом   
3765            г Уфа, ул. Нежинская, д.6  Многоквартирный дом   
3766            г Уфа, ул. Нежинская, д.6  Многоквартирный дом   
4320       г Уфа, ул. Революционная, д.88  Многоквартирный дом   
4321       г Уфа, ул. Революционная, д.88  Многоквартирный дом   
5727          г Уфа, ул. Энтузиастов, д.6  Многоквартирный дом   
5728          г Уфа, ул. Энтузиастов, д.6  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  \
1583                  9     1979-01-01                2022.00   
1584                  9     1979-01-01                8256.50   
1603                 12     1978-01-01                 655.00   
1604                 12     1978-01-01                3681.40   
2101                  1     1964-01-01                1161.90   
2102                  5     1964-01-01                2563.70   
2397                  1     1971-01-01                   0.01   
2398                  5     1971-01-01                3448.80   
3765                  9     1972-01-01                 427.70   
3766                  9     1972-01-01                2544.60   
4320                  1     1974-01-01                1381.20   
4321                  5     1974-01-01                2800.40   
5727                  9     1986-01-01                 979.10   
5728                  9     1986-01-01               12006.70   

                                       Адрес объекта 2  
1583   г Уфа, ул. Вологодская, д.13, Подобъект №984990  
1584                      г Уфа, ул. Вологодская, д.13  
1603               г Уфа, ул. Вологодская, д.20, extra  
1604                      г Уфа, ул. Вологодская, д.20  
2101        г Уфа, ул. Интернациональная, д.113, extra  
2102               г Уфа, ул. Интернациональная, д.113  
2397         г Уфа, ул. Кирова, д.95, Подобъект №46590  
2398                           г Уфа, ул. Кирова, д.95  
3765                  г Уфа, ул. Нежинская, д.6, extra  
3766                         г Уфа, ул. Нежинская, д.6  
4320  г Уфа, ул. Революционная, д.88, Подобъект №46372  
4321                    г Уфа, ул. Революционная, д.88  
5727    г Уфа, ул. Энтузиастов, д.6, Подобъект №984984  
5728                       г Уфа, ул. Энтузиастов, д.6

#### 2.1.3 Объекты типа ```Частный дом```

In [18]:
s = "Частный дом"
selection = df[df["Тип Объекта"] == s]["Адрес объекта"].tolist()
n_address = len(selection)

cond_1a = data["Тип объекта"]==s
cond_2a = data["Адрес объекта"].isin(selection)
cond_1b = buildings["Тип Объекта"]==s
cond_2b = buildings["Адрес объекта"].isin(selection)

df1 = data[cond_1a & cond_2a]
df2 = buildings[cond_1b & cond_2b]

n_objects = df2.shape[0]
n_billed_address = df1["Адрес объекта"].nunique()
n_odpu = df1["№ ОДПУ"].nunique()
consumed = df1["Текущее потребление, Гкал"].sum()

no_bills = set(selection) - set(df1["Адрес объекта"])
no_bills_objects = df2[df2['Адрес объекта'].isin(no_bills)].iloc[:, :-1]

print(f"На {n_address} неуникальных адресах находится {n_objects} объектов типа {s}.")
print(f'Общая площадь этих объектов:{df2["Общая площадь объекта"].sum(): .0f}.')
print(f'На {len(no_bills)} неуникальных адресах ({len(no_bills_objects)} объектов типа {s}) показаний ОДПУ нет.\n')

no_bills_objects

На 4 неуникальных адресах находится 12 объектов типа Частный дом.
Общая площадь этих объектов: 311.
На 4 неуникальных адресах (12 объектов типа Частный дом) показаний ОДПУ нет.



Адрес объекта  Тип Объекта  Этажность объекта  \
882    г Уфа, ул. 9 Января, д.12  Частный дом                  1   
883    г Уфа, ул. 9 Января, д.12  Частный дом                  1   
884    г Уфа, ул. 9 Января, д.12  Частный дом                  1   
885    г Уфа, ул. 9 Января, д.12  Частный дом                  1   
886    г Уфа, ул. 9 Января, д.12  Частный дом                  1   
1835     г Уфа, ул. Губкина, д.6  Частный дом                  1   
1836     г Уфа, ул. Губкина, д.6  Частный дом                  1   
4783    г Уфа, ул. Садовая, д.17  Частный дом                  1   
4784    г Уфа, ул. Садовая, д.17  Частный дом                  1   
5561  г Уфа, ул. Чукотская, д.23  Частный дом                  1   
5562  г Уфа, ул. Чукотская, д.23  Частный дом                  1   
5563  г Уфа, ул. Чукотская, д.23  Частный дом                  1   

     Дата постройки  Общая площадь объекта  
882      1998-01-01                  12.00  
883      1998-01-01                   9.60  
884      1998-01-01                  12.60  
885      1998-01-01                  16.23  
886      1998-01-01                  17.03  
1835     1989-01-01                  18.30  
1836     1989-01-01                  33.06  
4783     1947-01-01                  39.60  
4784     1947-01-01                  57.80  
5561     1989-01-01                  22.20  
5562     1989-01-01                  27.40  
5563     1989-01-01                  44.80

#### 2.1.4 Объекты типа ```Учебное заведение, комбинат, центр```

In [19]:
s = "Учебное заведение, комбинат, центр"
selection = df[df["Тип Объекта"] == s]["Адрес объекта"].tolist()
n_address = len(selection)

cond_1a = data["Тип объекта"]==s
cond_2a = data["Адрес объекта"].isin(selection)
cond_1b = buildings["Тип Объекта"]==s
cond_2b = buildings["Адрес объекта"].isin(selection)

df1 = data[cond_1a & cond_2a]
df2 = buildings[cond_1b & cond_2b]

n_objects = df2.shape[0]
n_billed_address = df1["Адрес объекта"].nunique()
n_odpu = df1["№ ОДПУ"].nunique()
consumed = df1["Текущее потребление, Гкал"].sum()

print(f"На {n_address} неуникальном адресе находится {n_objects} объекта типа {s}.")
print(f'Общая площадь этих объектов: {df2["Общая площадь объекта"].sum()}.')
print(f'На {n_billed_address} адрес приходится {n_odpu} № ОДПУ')
print(f'Oбщий объем потребления составил{consumed: .0f} Гкал.')

df2

На 1 неуникальном адресе находится 3 объекта типа Учебное заведение, комбинат, центр.
Общая площадь этих объектов: 7234.91.
На 1 адрес приходится 2 № ОДПУ
Oбщий объем потребления составил 4982 Гкал.


Адрес объекта                         Тип Объекта  \
4394  г Уфа, ул. Рихарда Зорге, д.27  Учебное заведение, комбинат, центр   
4395  г Уфа, ул. Рихарда Зорге, д.27  Учебное заведение, комбинат, центр   
4396  г Уфа, ул. Рихарда Зорге, д.27  Учебное заведение, комбинат, центр   

      Этажность объекта Дата постройки  Общая площадь объекта  \
4394                  1     1985-01-01                   0.01   
4395                  5     1985-01-01                4008.00   
4396                 12     1985-01-01                3226.90   

                                      Адрес объекта 2  
4394                   г Уфа, ул. Рихарда Зорге, д.27  
4395  г Уфа, ул. Рихарда Зорге, д.27, № ОДПУ 00111382  
4396  г Уфа, ул. Рихарда Зорге, д.27, № ОДПУ 00102590

1. На 46 неуникальных адресах находится 106 объектов типа ```Другое строение```.
  * Общая площадь этих объектов не учитывается (2.62).
  * На 34 адреса из 46 приходится 66 ```№ ОДПУ```. Oбщий объем потребления составил 75883 Гкал.
  * На 12 неуникальных адресах (29 объектов типа Другое строение) данных показаний учёта тепловой энергии нет.

2. На 7 неуникальных адресах находится 14 объектов типа ```Многоквартирный дом``` - по 2 на каждый адрес:
  * ```г Уфа, ул. Вологодская, д.13```,
  * ```г Уфа, ул. Вологодская, д.20```,
  * ```г Уфа, ул. Интернациональная, д.113```,
  * ```г Уфа, ул. Кирова, д.95```,
  * ```г Уфа, ул. Нежинская, д.6```,
  * ```г Уфа, ул. Революционная, д.88```,
  * ```г Уфа, ул. Энтузиастов, д.6```.

  * Общая площадь этих объектов - 41929, oбщий объем потребления - 9436 Гкал.
  * Адреса многоквартирных домов сделаны уникальными в признаке ```Адрес объекта 2``` через сопоставление с данными об учете теплоэнергии.
  * На трех объектах ниже счета не выставляются, для них в признаке ```Адрес объекта 2``` к адресу добавлено слово ```extra``` :
    * ```г Уфа, ул. Вологодская, д.20```,
    * ```г Уфа, ул. Интернациональная, д.113```,
    * ```г Уфа, ул. Нежинская, д.6```.

3. На 4 неуникальных адресах находится 12 объектов типа ```Частный дом``.
  * Общая площадь этих объектов: 311.
  * Ни у одного из 12 объектов данных показаний учёта тепловой энергии нет.
4. На 1 неуникальном адресе находится 3 объекта типа ```Учебное заведение, комбинат, центр	``` .
  * Общая площадь этих объектов: 7234.91.
  * На 1 адрес приходится 2 № ОДПУ
  * Oбщий объем потребления составил 4982 Гкал.
  * Адреса сделаны уникальными в признаке ```Адрес объекта 2``` через сопоставление с данными об учете теплоэнергии и проставление ``№ ОДПУ`` в адрес.

### 2.2. Объекты с Общей площадью <= 0.1

In [20]:
cond1 = buildings["Тип Объекта"] == "Многоквартирный дом"
cond2 = buildings["Общая площадь объекта"] <= 0.1

df = buildings[cond1 & cond2].iloc[:, :-1]
print(f"Число объектов без данных о площади: {df.shape[0]}")

df.head(2)

Число объектов без данных о площади: 25


Адрес объекта          Тип Объекта  \
949                      г Уфа, ул. Айская  Многоквартирный дом   
1452  г Уфа, ул. Большая Гражданская, д.49  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  
949                  20            NaT                   0.01  
1452                  1     2018-11-01                   0.02

* У 25 многоквартирных домов Общая площадь объекта не превышает 0.1.

### 2.3. Объекты без Даты постройки

In [21]:
cond3 = buildings["Дата постройки"].isnull()
df = buildings[cond1 & cond3].iloc[:, :-1]

print(f"Число объектов без данных о дате постройки: {df.shape[0]}")

df.head(2)

Число объектов без данных о дате постройки: 34


Адрес объекта          Тип Объекта  \
62       г Уфа, б-р. Ибрагимова, д.90 корп.2  Многоквартирный дом   
836  г Уфа, ул. 50-летия Октября, д.3 корп.П  Многоквартирный дом   

     Этажность объекта Дата постройки  Общая площадь объекта  
62                  20            NaT                53759.1  
836                  5            NaT                  279.6

У 34 многоквартирных дома нет признака Дата постройки.

### 2.4. Подобъекты

In [22]:
cond1_a = data["Тип объекта"] == "Многоквартирный дом"
cond1_b = buildings["Тип Объекта"] == "Многоквартирный дом"

addr = sorted(list(set(data[cond1_a]["Адрес объекта"]) - set(buildings[cond1_b]["Адрес объекта"])))
addr

['г Уфа, б-р. Давлеткильдеева, д.18, Подобъект №987152',
 'г Уфа, б-р. Давлеткильдеева, д.5, Подобъект №985759',
 'г Уфа, б-р. Ибрагимова, д.34, Подобъект №983365',
 'г Уфа, б-р. Молодежный, д.2, Подобъект №992132',
 'г Уфа, б-р. Молодежный, д.6 корп.1, Подобъект №983457',
 'г Уфа, б-р. Тухвата Янаби, д.22, Подобъект №351681',
 'г Уфа, пр-кт Октября, д.122, Подобъект №42688',
 'г Уфа, пр-кт Октября, д.174/2, Подобъект №985197',
 'г Уфа, пр-кт Октября, д.25, Подобъект №48042',
 'г Уфа, пр-кт Октября, д.25, Подобъект №48043',
 'г Уфа, пр-кт Октября, д.25, Подобъект №48044',
 'г Уфа, пр-кт Октября, д.60 корп.1, Подобъект №51211',
 'г Уфа, пр-кт Октября, д.78 корп.3, Подобъект №989097',
 'г Уфа, ул. 50 лет СССР, д.2, Подобъект №47840',
 'г Уфа, ул. 8 Марта, д.12 корп.3, Подобъект №49931',
 'г Уфа, ул. Академика Королева, д.2 корп.2, Подобъект №51212',
 'г Уфа, ул. Ахметова, д.324, Подобъект №986621',
 'г Уфа, ул. Баязита Бикбая, д.44, Подобъект №983215',
 'г Уфа, ул. Блюхера, д.17, Подобъе

In [23]:
len(addr)

77

In [24]:
s = 'г Уфа, ул. Энтузиастов, д.14'
cond2_b = buildings["Адрес объекта"].str.match(s)
cond2_a = data["Адрес объекта"].str.match(s)

print(f"{data[cond1_a & cond2_a]['Адрес объекта'].unique()}\n")
buildings[cond1_b & cond2_b].iloc[:, :-1]

['г Уфа, ул. Энтузиастов, д.14'
 'г Уфа, ул. Энтузиастов, д.14, Подобъект №984513'
 'г Уфа, ул. Энтузиастов, д.14, Подобъект №984848'
 'г Уфа, ул. Энтузиастов, д.14, Подобъект №984765']



Адрес объекта          Тип Объекта  Этажность объекта  \
5715  г Уфа, ул. Энтузиастов, д.14  Многоквартирный дом                 18   

     Дата постройки  Общая площадь объекта  
5715     2018-01-01                27686.6

In [25]:
data[data["Адрес объекта"]=="г Уфа, ул. Энтузиастов, д.14"]["№ ОДПУ"].unique()

array(['278553,278511'], dtype=object)

In [26]:
data[data["Адрес объекта"]=="г Уфа, ул. Энтузиастов, д.14, Подобъект №984765"]["№ ОДПУ"].unique()

array(['18480739'], dtype=object)

* У подобъектов и объектов отличается ```№ ОДПУ```- учет ведется раздельно, при этом площадь - общая.
* Для использования в моделях необходимо включить подобъекты в объекты, при выдаче результата - оценить площадь подобъеков и объектов в разрезе ```№ ОДПУ```.

In [27]:
sorted(list(set(data[cond1_a]["Адрес объекта 2"]) - set(buildings[cond1_b]["Адрес объекта 2"])))

['г Уфа, ул. Пархоменко, д.156 корп.Литер 2в',
 'г Уфа, ул. Уфимское шоссе, д.4']

### 2.5. Многоквартирные дома, отсутствующие в базе данных объектов

#### 2.5.1 ```г Уфа, ул. Пархоменко, д.156 корп.Литер 2в```

In [28]:
s = 'г Уфа, ул. Пархоменко, д.156'
cond2_b = buildings["Адрес объекта"].str.match(s)
cond2_a = data["Адрес объекта"].str.match(s)

print(f"{data[cond1_a & cond2_a]['Адрес объекта'].unique()}\n")
buildings[cond1_b & cond2_b].iloc[:, :-1]

['г Уфа, ул. Пархоменко, д.156 корп.Литер 2в'
 'г Уфа, ул. Пархоменко, д.156 корп.б'
 'г Уфа, ул. Пархоменко, д.156 корп.в']



Адрес объекта          Тип Объекта  \
3937  г Уфа, ул. Пархоменко, д.156 корп.А  Многоквартирный дом   
3938  г Уфа, ул. Пархоменко, д.156 корп.б  Многоквартирный дом   
3939  г Уфа, ул. Пархоменко, д.156 корп.в  Многоквартирный дом   
3940  г Уфа, ул. Пархоменко, д.156 корп.г  Многоквартирный дом   
3941  г Уфа, ул. Пархоменко, д.156 корп.Д  Многоквартирный дом   
3942  г Уфа, ул. Пархоменко, д.156 корп.е  Многоквартирный дом   
3943  г Уфа, ул. Пархоменко, д.156 корп.ж  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  
3937                 25     2021-11-01               10428.70  
3938                 16     2019-11-01               21415.40  
3939                 25     2019-11-01               11241.00  
3940                 25     2022-11-01               10515.70  
3941                 16     2021-11-01                9040.71  
3942                 25            NaT                   0.01  
3943                 25            NaT                   0.10

* ```г Уфа, ул. Пархоменко, д.156 корп.Литер 2в``` в ```data``` не соответствует
аналогу в ```buildings```:
  * ```г Уфа, ул. Пархоменко, д.156 корп.А```
  * ```г Уфа, ул. Пархоменко, д.156 корп.г```
  * ```г Уфа, ул. Пархоменко, д.156 корп.Д```

#### 2.5.2 ```г Уфа, ул. Уфимское шоссе, д.4```

In [29]:
s = 'г Уфа, ул. Уфимское шоссе, д.4'
cond2_b = buildings["Адрес объекта"].str.match(s)
cond2_a = data["Адрес объекта"].str.match(s)

print(f"{data[cond1_a & cond2_a]['Адрес объекта'].unique()}\n")
buildings[cond1_b & cond2_b].iloc[:, :-1]

['г Уфа, ул. Уфимское шоссе, д.4' 'г Уфа, ул. Уфимское шоссе, д.4 корп.1']



Адрес объекта          Тип Объекта  \
5275  г Уфа, ул. Уфимское шоссе, д.4 корп.1  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  
5275                  9     1976-01-01                 3821.3

In [30]:
data[data["Адрес объекта"]=="г Уфа, ул. Уфимское шоссе, д.4"]["№ ОДПУ"].unique()

array(['102522'], dtype=object)

In [31]:
data[data["Адрес объекта"]=="г Уфа, ул. Уфимское шоссе, д.4 корп.1"]["№ ОДПУ"].unique()

array(['104237'], dtype=object)

* Объекта ```г Уфа, ул. Уфимское шоссе, д.4``` нет. Есть только ```г Уфа, ул. Уфимское шоссе, д.4 корп.1```, но у него другой ```№ ОДПУ```.

#### 2.5.3. ```г Уфа, ул. Владивостокская, д.10 корп.3,4```

In [32]:
s = 'г Уфа, ул. Владивостокская, д.10'
cond2_b = buildings["Адрес объекта"].str.match(s)
cond2_a = data["Адрес объекта"].str.match(s)

print(f"{data[cond1_a & cond2_a]['Адрес объекта'].unique()}\n")
buildings[cond1_b & cond2_b].iloc[:, :-1]

['г Уфа, ул. Владивостокская, д.10'
 'г Уфа, ул. Владивостокская, д.10 корп.3,4']



Адрес объекта          Тип Объекта  \
1554  г Уфа, ул. Владивостокская, д.10  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  
1554                 14     2015-01-01               23745.65

In [33]:
data[data["Адрес объекта"]=="г Уфа, ул. Владивостокская, д.10"]["№ ОДПУ"].unique()

array(['248788/15024285/20101973/23-190725', '20-101973'], dtype=object)

In [34]:
data[data["Адрес объекта"]=="г Уфа, ул. Владивостокская, д.10 корп.3,4"]["№ ОДПУ"].unique()

array(['15024285'], dtype=object)

In [35]:
data[data["Адрес объекта"]=="г Уфа, ул. Владивостокская, д.10 корп.3,4"].iloc[:, :-1]

index Подразделение    № ОДПУ Вид энерг-а ГВС  \
98        98           Уфа  15024285             NaN   
8764       0           Уфа  15024285             NaN   
11086     22           Уфа  15024285             NaN   
14343     61           Уфа  15024285             NaN   
23086     22           Уфа  15024285             NaN   
36211     97           Уфа  15024285             NaN   
40445     17           Уфа  15024285             NaN   
45977     87           Уфа  15024285             NaN   
46971     35           Уфа  15024285             NaN   
48359    118           Уфа  15024285             NaN   
52719    139           Уфа  15024285             NaN   
61122     55           Уфа  15024285             NaN   
67824   1277           Уфа  15024285             NaN   

                                   Адрес объекта          Тип объекта  \
98     г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
8764   г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
11086  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
14343  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
23086  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
36211  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
40445  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
45977  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
46971  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
48359  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
52719  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
61122  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   
67824  г Уфа, ул. Владивостокская, д.10 корп.3,4  Многоквартирный дом   

      Дата текущего показания  Текущее потребление, Гкал Период потребления  
98                        NaT                        NaN         2022-04-01  
8764                      NaT                        NaN         2022-06-01  
11086                     NaT                        NaN         2021-09-01  
14343              2022-03-01                    221.453         2022-02-01  
23086                     NaT                        NaN         2022-01-01  
36211                     NaT                        NaN         2021-11-01  
40445                     NaT                        NaN         2021-08-01  
45977                     NaT                        NaN         2021-07-01  
46971                     NaT                        NaN         2022-05-01  
48359                     NaT                        NaN         2021-12-01  
52719                     NaT                        NaN         2021-10-01  
61122                     NaT                        NaN         2022-03-01  
67824                     NaT                        NaN         2023-02-01

* Объект ```г Уфа, ул. Владивостокская, д.10 корп.3,4``` в data должен попасть в ```г Уфа, ул. Владивостокская, д.10``` - корп.3,4 избыточен.

#### 2.5.4 ``` г Уфа, ул. Даута Юлтыя, д.12```

In [36]:
s = 'г Уфа, ул. Даута Юлтыя, д.12'
cond2_b = buildings["Адрес объекта"].str.match(s)
cond2_a = data["Адрес объекта"].str.match(s)

print(f"{data[cond1_a & cond2_a]['Адрес объекта'].unique()}\n")
buildings[cond1_b & cond2_b].iloc[:, :-1]

['г Уфа, ул. Даута Юлтыя, д.12, Подъезд №20654'
 'г Уфа, ул. Даута Юлтыя, д.12, Подъезд №20655'
 'г Уфа, ул. Даута Юлтыя, д.12']



Адрес объекта          Тип Объекта  Этажность объекта  \
1841  г Уфа, ул. Даута Юлтыя, д.12  Многоквартирный дом                 16   

     Дата постройки  Общая площадь объекта  
1841     2011-01-01                13465.5

In [37]:
data[data["Адрес объекта"]=="г Уфа, ул. Даута Юлтыя, д.12"]["№ ОДПУ"].unique()

array(['89612 89617'], dtype=object)

In [38]:
data[data["Адрес объекта"]=="г Уфа, ул. Даута Юлтыя, д.12, Подъезд №20654"]["№ ОДПУ"].unique()

array(['89612'], dtype=object)

In [39]:
data[data["Адрес объекта"]=="г Уфа, ул. Даута Юлтыя, д.12, Подъезд №20655"]["№ ОДПУ"].unique()

array(['89617'], dtype=object)

* В объекте ```г Уфа, ул. Даута Юлтыя, д.12``` подобъекты следует объединить - разделены несистемно:
  * ```г Уфа, ул. Даута Юлтыя, д.12```,
  * ```г Уфа, ул. Даута Юлтыя, д.12, Подъезд №20654```
  * ```г Уфа, ул. Даута Юлтыя, д.12, Подъезд №20655```

#### 2.5.5 ```г Уфа, ул. Комсомольская, д.15..```

In [40]:
s = 'г Уфа, ул. Комсомольская, д.15'
cond2_b = buildings["Адрес объекта"].str.match(s)
cond2_a = data["Адрес объекта"].str.match(s)

print(f"{data[cond1_a & cond2_a]['Адрес объекта'].unique()}\n")
buildings[cond1_b & cond2_b].iloc[:, :-1]

['г Уфа, ул. Комсомольская, д.159 корп.1'
 'г Уфа, ул. Комсомольская, д.153 корп.1'
 'г Уфа, ул. Комсомольская, д.159' 'г Уфа, ул. Комсомольская, д.15'
 'г Уфа, ул. Комсомольская, д.153' 'г Уфа, ул. Комсомольская, д.157'
 'г Уфа, ул. Комсомольская, д.157 корп.1'
 'г Уфа, ул. Комсомольская, д.15, Подобъект №35951'
 'г Уфа, ул. Комсомольская, д.15..']



Адрес объекта          Тип Объекта  \
2780          г Уфа, ул. Комсомольская, д.15  Многоквартирный дом   
2781         г Уфа, ул. Комсомольская, д.153  Многоквартирный дом   
2782  г Уфа, ул. Комсомольская, д.153 корп.1  Многоквартирный дом   
2784         г Уфа, ул. Комсомольская, д.157  Многоквартирный дом   
2785  г Уфа, ул. Комсомольская, д.157 корп.1  Многоквартирный дом   
2786         г Уфа, ул. Комсомольская, д.159  Многоквартирный дом   
2787  г Уфа, ул. Комсомольская, д.159 корп.1  Многоквартирный дом   

      Этажность объекта Дата постройки  Общая площадь объекта  
2780                 23     2013-01-01                59310.4  
2781                  9     1976-01-01                 7698.9  
2782                 12     1982-01-01                 4794.0  
2784                 12     1957-01-01                 4758.4  
2785                  9     1976-01-01                 3830.4  
2786                  9     1976-01-01                10843.5  
2787                  7     2005-01-01                15365.5

In [41]:
data[data["Адрес объекта"]=="г Уфа, ул. Комсомольская, д.15.."]

index Подразделение  № ОДПУ Вид энерг-а ГВС  \
66028    598           Уфа  224503             NaN   

                          Адрес объекта          Тип объекта  \
66028  г Уфа, ул. Комсомольская, д.15..  Многоквартирный дом   

      Дата текущего показания  Текущее потребление, Гкал Период потребления  \
66028                     NaT                        NaN         2022-08-01   

                      Адрес объекта 2  
66028  г Уфа, ул. Комсомольская, д.15

* Объекта ```г Уфа, ул. Комсомольская, д.15..``` нет. Адрес объекта необходимо исправить.

### 2.6. Объекты без данных о выставленных счетах

In [42]:
addr_2 = sorted(
    list(set(buildings[cond1_b]["Адрес объекта 2"]) - set(data[cond1_a]["Адрес объекта 2"])))
df = buildings[cond1_b & buildings["Адрес объекта 2"].isin(addr_2)].iloc[:, :-1]
n = df.shape[0]
s = df['Общая площадь объекта'].sum() / 1e06
print(f"У {n} объектов общей площадью {s: 0.1f} млн. нет данных о выствленных счетах.")

save = False
if save:
  path = "/content/drive/MyDrive/ML_projects/08_anomaly_detection/data/01_raw/"
  file_name = "Тип строения, этажность, площадь, год постройки 2.xlsx"
  df.to_excel(f"{path}{file_name}")

У 752 объектов общей площадью  1.1 млн. нет данных о выствленных счетах.


## 3. Комментарии по пропускам в данных

1. В данных по объектам 58 неуникальных адресов:

  1.1. На 46 неуникальных адресах находится 106 объектов типа ```Другое строение```.
    * Общая площадь этих объектов не учитывается (2.62).
    * На 34 адреса из 46 приходится 66 ```№ ОДПУ```. Oбщий объем потребления составил 75883 Гкал.
    * На 12 неуникальных адресах (29 объектов типа Другое строение) данных показаний учёта тепловой энергии нет.

  1.2. На 7 неуникальных адресах находится 14 объектов типа ```Многоквартирный дом``` - по 2 на каждый адрес:
    * ```г Уфа, ул. Вологодская, д.13```,
    * ```г Уфа, ул. Вологодская, д.20```,
    * ```г Уфа, ул. Интернациональная, д.113```,
    * ```г Уфа, ул. Кирова, д.95```,
    * ```г Уфа, ул. Нежинская, д.6```,
    * ```г Уфа, ул. Революционная, д.88```,
    * ```г Уфа, ул. Энтузиастов, д.6```.

    * Общая площадь этих объектов - 41929, oбщий объем потребления - 9436 Гкал.
    * Адреса многоквартирных домов сделаны уникальными в признаке ```Адрес объекта 2``` через сопоставление с данными об учете теплоэнергии.
    * На трех объектах ниже счета не выставляются, для них в признаке ```Адрес объекта 2``` к адресу добавлено слово ```extra``` :
      * ```г Уфа, ул. Вологодская, д.20```,
      * ```г Уфа, ул. Интернациональная, д.113```,
      * ```г Уфа, ул. Нежинская, д.6```.

  1.3. На 4 неуникальных адресах находится 12 объектов типа ```Частный дом``.
    * Общая площадь этих объектов: 311.
    * Ни у одного из 12 объектов данных показаний учёта тепловой энергии нет.
  1.4. На 1 неуникальном адресе находится 3 объекта типа ```Учебное заведение, комбинат, центр	``` .
    * Общая площадь этих объектов: 7234.91.
    * На 1 адрес приходится 2 № ОДПУ
    * Oбщий объем потребления составил 4982 Гкал.
    * Адреса сделаны уникальными в признаке ```Адрес объекта 2``` через сопоставление с данными об учете теплоэнергии и проставление ``№ ОДПУ`` в адрес.

2. У 25 многоквартирных домов Общая площадь объекта не превышает 0.1.

3. У 34 многоквартирных дома нет признака Дата постройки.

4. У подобъектов и объектов отличается ```№ ОДПУ```- учет ведется раздельно, при этом площадь - общая.
    *  Для использования в моделях необходимо включить подобъекты в объекты, при выдаче результата - оценить площадь подобъеков и объектов в разрезе ```№ ОДПУ```.

5. Многоквартирные дома, отсутствующие в базе данных объектов:
    * ```г Уфа, ул. Пархоменко, д.156 корп.Литер 2в``` в ```data``` не соответствует
    аналогу в ```buildings```:
      * ```г Уфа, ул. Пархоменко, д.156 корп.А```
      * ```г Уфа, ул. Пархоменко, д.156 корп.г```
      * ```г Уфа, ул. Пархоменко, д.156 корп.Д```
    * Объекта ```г Уфа, ул. Уфимское шоссе, д.4``` нет. Есть только ```г Уфа, ул. Уфимское шоссе, д.4 корп.1```, но у него другой ```№ ОДПУ```.
    * Объект ```г Уфа, ул. Владивостокская, д.10 корп.3,4``` в data должен попасть в ```г Уфа, ул. Владивостокская, д.10``` -  ```корп.3,4``` избыточен.
    * В объекте ```г Уфа, ул. Даута Юлтыя, д.12``` подобъекты следует объединить - разделены несистемно:
      * ```г Уфа, ул. Даута Юлтыя, д.12```,
      * ```г Уфа, ул. Даута Юлтыя, д.12, Подъезд №20654```
      * ```г Уфа, ул. Даута Юлтыя, д.12, Подъезд №20655```
    * Объекта ```г Уфа, ул. Комсомольская, д.15..``` нет. Адрес объекта необходимо исправить.

6. У 752 объектов общей площадью  1.1 млн. нет данных о выствленных счетах.